File: go/69. Sqrt(x).go
func mySqrt(x int) int {
	if x < 2 {
		return x
	}
	start, end := 1, x>>1
	for start < end {
		mid := int((start + end + 1) >> 1)
		if mid <= x/mid {
			start = mid
		} else {
			end = mid - 1
		}
	}
	return end
}

File: 69. Sqrt(x).py
class Solution:
    def mySqrt(self, x: int) -> int:
        start = 0
        end = x
        while start <= end:
            mid = (start + end) >> 1
            if mid*mid <= x and (mid+1)*(mid+1) > x:
                return mid
            if mid * mid < x:
                start = mid+1
            else:
                end = mid - 1
        return end
